### PR TITLE
pythonPackages.imgaug: 0.3.0-patch

### PR DESCRIPTION
**Pull request by Rakesh4G**
Added patch to bypass opencv-python dependency from imgaug

@CMCDragonkai

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @CMCDragonkai @FRidh

### CODE REVIEW

**Comment by Rakesh4G** [comment]
package has been updated as per inputs from 
[https://github.com/NixOS/nixpkgs/issues/61936](url)
[https://github.com/aleju/imgaug/issues/379#issuecomment-520987731](url)

**Comment by CMCDragonkai** [comment]
Note that in the master branch of imgaug, the author has changed to `opencv-python-headless` from `opencv-python`. See: https://github.com/aleju/imgaug/blob/master/setup.py#L21.

This means for this release 0.2.9, it should work, but the next release may require a change to the substitute patch.

**Review by CMCDragonkai — APPROVED** [review]
Approved, looks good. Mark it ready for review.

**Comment by mmahut** [comment]
@GrahamcOfBorg build python37Packages.imgaug

**Comment by mmahut** [comment]
Looks like resulting derivation is empty?

```
[nix-shell:~/.cache/nix-review/pr-67494/results/python37Packages.imgaug]# tree
.
└── nix-support
    └── propagated-build-inputs

1 directory, 1 file
```

**Comment by CMCDragonkai** [comment]
This is blocked on https://github.com/aleju/imgaug/issues/414

File: pkgs/development/python-modules/imgaug/default.nix
Comment (Rakesh4G) [reply pkgs/development/python-modules/imgaug/default.nix]: > I think should be `opencv-python-headless`

Hi @tbenst, currently i am using imgaug version `0.2.9` which uses ` opencv-python `.
 `opencv-python-headless` is used in imgaug `0.3.0` onwards.

Comment (tbenst) [reply pkgs/development/python-modules/imgaug/default.nix]: @Rakesh4G master is now on 0.3.0: https://github.com/NixOS/nixpkgs/commit/598db503ecdfbfa1ba400fef26edc5b67efb204b

Comment (Rakesh4G) [reply pkgs/development/python-modules/imgaug/default.nix]: @tbenst Ok, I will update this to 0.3.0 as well.

Comment (Rakesh4G) [reply pkgs/development/python-modules/imgaug/default.nix]: Done.

**Comment by jonringer** [comment]
do you mind rebasing on top of master? apparently there's some merge conflicts:
```
git pull -r origin master
git push <fork> <branch> --force
```

**Comment by Rakesh4G** [comment]
> do you mind rebasing on top of master? apparently there's some merge conflicts:
> 
> ```
> git pull -r origin master
> git push <fork> <branch> --force
> ```
 Sure, I will.

**Comment by tbenst** [comment]
Given that imgaug is functioning on waiting on upstream for a fix, perhaps we should merge & fix broken package?

**Comment by Rakesh4G** [comment]
The tests hang issue with imgaug-0.3.0 has been fixed on master. I have taken the latest changes and  current expression is ready to be merged.
@tbenst @CMCDragonkai @jonringer

**Comment by CMCDragonkai** [comment]
When attempting to install the latest changes using `nix-env -f ./default.nix -iA 'pkgs.python3Packages.imgaug`, the tests for hang here:

```
test/augmenters/test_meta.py::TestAugmenter_augment_batches::test_augment_batches_list_of_list_of_invalid_datatype PASSED
```

The build just stops there, the `pytest` command is still running but with 0% CPU usage.

**Comment by Rakesh4G** [comment]
> nix-env -f ./default.nix -iA 'pkgs.python3Packages.imgaug

yup, you are correct. I am also facing this issue with above command you mentioned.
looks like i had missed something. I will re-inspect and see what is going wrong.

**Comment by Rakesh4G** [comment]
I have updated the error details and other findings on https://github.com/aleju/imgaug/issues/414

**Comment by Rakesh4G** [comment]
This PR is ready for merge. All the tests are now passing.
@tbenst @jonringer @CMCDragonkai

**Review by jonringer — APPROVED** [review]
diff LGTM
commit LGTM
tests pass :+1: 

```
[3 built, 0.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/67494
1 package were built:
python37Packages.imgaug
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build python37Packages.imgaug

File: pkgs/development/python-modules/imgaug/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/imgaug/default.nix]:
```suggestion
    maintainers = with maintainers; [ cmcdragonkai rakesh4g ];
    platforms = platforms.linux;
```
looks like darwin has odd behavior:
```
Current thread 0x0000700003d9d000 (most recent call first):
  File "/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/augmenters/blur.py", line 262 in blur_gaussian_
  File "/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/augmenters/blur.py", line 466 in _augment_batch
  File "/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/augmenters/meta.py", line 606 in augment_batch
  File "/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/multicore.py", line 506 in _Pool_worker
  File "/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99Fatal Python error: Segmentation fault
```

File: pkgs/development/python-modules/imgaug/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/imgaug/default.nix]: ${python.interpreter}

File: pkgs/development/python-modules/imgaug/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/imgaug/default.nix]: space between comma and parameter

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/imgaug/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/imgaug/default.nix]: or just
```suggestion
     pytest ./test
```

**Comment by CMCDragonkai** [comment]
I believe those Darwin problems are the same as NixOS problems until imgaug maintainers disabled those tests on NixOS platform. But not on Mac.

On 20 December 2019 06:08:30 GMT+11:00, Jon <notifications@github.com> wrote:
>jonringer commented on this pull request.
>
>
>
>>  
>   meta = with stdenv.lib; {
>     homepage = https://github.com/aleju/imgaug;
>   description = "Image augmentation for machine learning experiments";
>     license = licenses.mit;
>-    maintainers = with maintainers; [ cmcdragonkai ];
>-    broken = true; # opencv-python bindings aren't available yet, and
>look non-trivial
>+    maintainers = with maintainers; [ cmcdragonkai rakesh4g ];
>
>```suggestion
>    maintainers = with maintainers; [ cmcdragonkai rakesh4g ];
>    platforms = platforms.linux;
>```
>looks like darwin has odd behavior:
>```
>Current thread 0x0000700003d9d000 (most recent call first):
>File
>"/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/augmenters/blur.py",
>line 262 in blur_gaussian_
>File
>"/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/augmenters/blur.py",
>line 466 in _augment_batch
>File
>"/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/augmenters/meta.py",
>line 606 in augment_batch
>File
>"/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99a420efc45652a1264920dc20378a54b1325/imgaug/multicore.py",
>line 506 in _Pool_worker
>File
>"/private/tmp/nix-build-python3.7-imgaug-0.3.0.drv-0/imgaug-c3d99Fatal
>Python error: Segmentation fault
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/67494#pullrequestreview-334868800

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by CMCDragonkai** [comment]
We don't have Mac platforms to test this. So we could forward on the details to imgaug maintainer. While also just only enabling this for Linux?

**Comment by CMCDragonkai** [comment]
Another idea is a patch in Nix package level that disables those specific tests.

**Comment by jonringer** [comment]
personally, I would either do `doCheck = stdenv.isLinux` or `meta.platforms = platforms.linux`. The meta platform is probably more appropriate if we don't have some other form of testing.

File: pkgs/development/python-modules/imgaug/default.nix
Comment (Rakesh4G) [reply pkgs/development/python-modules/imgaug/default.nix]: Done.

File: pkgs/development/python-modules/imgaug/default.nix
Comment (Rakesh4G) [reply pkgs/development/python-modules/imgaug/default.nix]: Done.

File: pkgs/development/python-modules/imgaug/default.nix
Comment (Rakesh4G) [reply pkgs/development/python-modules/imgaug/default.nix]: Done

**Comment by Rakesh4G** [comment]
I have updated the requested changes. Thanks.
@FRidh @jonringer @CMCDragonkai

**Comment by jonringer** [comment]
@GrahamcOfBorg build python37Packages.imgaug

**Comment by tbenst** [comment]
I cherry-picked this patch and has been working for me! would love to see merged

**Pull request merged**
